### DISCUSSION TITLE
How to flash a modified BIOS on ASUS Motherboards (Guide for NvStrapsReBar)

### DISCUSSION THREAD

**Discussion by H-Sami** [terminatorul/NvStrapsReBar#61] [Show and tell]
Hello,
I have a Z370-E motherboard and I've been having a lot of issues with flashing a modified BIOS with NvStrapsReBar.ffs
The first issue I've encountered was the Asus EZ flash bios rejecting the modified image file. This is due to the nature of the CAP file (Capsuled) and the fact that it's encrypted. I've tried other methods such as using Asus' AI-Suit software and manipulating the software/tool into thinking that it's flashing a non-modified BIOS file. The issue with that is after the BIOS is flashed, the NvStrapsReBar.ffs driver does not get loaded with the new bios. 

After a lot of digging, I've finally managed to find a workaround and flash a modified BIOS with NvStrapsReBar.ffs using AFUDOS and have successfully enabled ReBar on my RTX 2060.

The solution to this problem is very simple - I'll walk you through the steps of how you can flash the modified bios file.

First off, you'll need the following tools:
1) AFUDOS
2) Rufus
3) UEFI Tool
4) A fresh BIOS image from ASUS' website

Link: 
[BIOS TOOLS.zip](https://github.com/terminatorul/NvStrapsReBar/files/14620080/BIOS.TOOLS.zip)

Virustotal Scan: 4e8857723ac1cf5ef0a461725a9b8b197c5fd40af341f6a55a5ac9c1f56f7962

### **First Part: Prepping the USB drive**

Open up rufus and make sure you have the following settings toggled/enabled:

![image](https://github.com/terminatorul/NvStrapsReBar/assets/163582836/21f48fc6-8c1e-40a1-949f-e78d122986d6)

Then click on start. This will erase and format your USB drive and make it bootable.
Once done, navigate to the USB directory and then copy the files from "Copy to USB" to the USB's root directory. There should be 2 files, AFUDOS.exe and AUTOEXEC.bat. You will get promoted by Windows, just click on replace. 

### **Second Part: Prepping the BIOS using UEFI Tool**

Open up UEFI Tool and load the CAP bios image that you've downloaded from ASUS' website.

![image](https://github.com/terminatorul/NvStrapsReBar/assets/163582836/4fc6b539-18a2-4fe6-8755-0e368069d5f0)
Now each BIOS is going to be slightly different, what you're going to do is find the DXE driver region on your BIOS, you should eventually find the region that contains a lot of DXE drivers. Scroll down to the end of the region then right-click on last driver and click on "insert after" then select your NvStrapsReBar.ffs file.

![image](https://github.com/terminatorul/NvStrapsReBar/assets/163582836/eb305bc8-4b17-443b-ae27-ac898883c421)
The end result should look like this:

![image](https://github.com/terminatorul/NvStrapsReBar/assets/163582836/f0d577bd-94e1-4be9-b100-324ff9e7af5d)

Save the image as a CAP file.

Reopen UEFI Tool and open up the modified CAP image. You can double-check if you still have the NvStrapsReBar.ffs driver. (You should have it)
Now here comes the trick, you will need right-click on AMI Aptio capsule at the top and select "Extract body" then save the bios file as a .rom file. This will essentially make it bootable using AFUDOS. Reopen the .rom file using UEFI Tool to double-check if you still have the NvStrapsReBar.ffs driver. Also, make sure the bios' file name is "bios.rom"

Now that you have your BIOS file ready, go ahead and copy it to your USB drive. You should have all these files on your USB drive:

![image](https://github.com/terminatorul/NvStrapsReBar/assets/163582836/5dc57e79-f557-47b0-b7d3-2510e20cc7ca)
_Optional step, but it's recommended:_
Load into your BIOS and "Load optimized defaults". Save it and restart your PC. 

### **Final Part: Flashing the modified BIOS**

Load into your BIOS and enable CSM, then navigate to the Boot Tab and scroll down till you find your Bootable devices. Select your USB drive _without_ UEFI mode and hit Enter. This will load AFUDOS and it will start flashing your new modified BIOS automatically.

![image](https://github.com/terminatorul/NvStrapsReBar/assets/163582836/1d63a196-cd59-46aa-a3b9-e6fbfeb2fb52)
Once it's done, restart your PC and go into your BIOS once more and Disable CSM and Enable Above 4G Decoding and Resizable Bar, then restart your PC. 

Now that you have your NvStrapsReBar.ffs loaded onto your BIOS, you will be able to use NvStrapsReBar.exe and enable ReBar.

### **Results:**
![image](https://github.com/terminatorul/NvStrapsReBar/assets/163582836/60c3e108-0437-4ea1-8658-03d9ca54671a)
![image](https://github.com/terminatorul/NvStrapsReBar/assets/163582836/62f19a74-350a-4f52-8bf0-f9c28ce139a2)

Thank you for reading. 

Regards,
Harveyyy

**Comment by terminatorul** [comment]
Thank you for this ! Looks like it will save many people

**Comment by Epikest** [comment]
doesn't work for me:
![PXL_20240316_163030390.jpg](https://github.com/terminatorul/NvStrapsReBar/assets/45499416/83fd2086-205d-4ee2-b9e9-acb4857bca37)

**Reply by H-Sami** [reply]
What's your motherboard?

**Reply by Epikest** [reply]
Asus ROG STRIX B550-F GAMING (WI-FI)

**Reply by manusalgam** [reply]
Exactly same error on TUF GAMING X570-PLUS (BIOS 5003), I've tested flashing 3405 (oldest BIOS wasn't able to flash through EZ Updater, maybe because it's not possible to downgrade from AGESA 1.2.X), but nothing changed. I've tested even corrupting bios (switching off power when upgrading) and loading mod by crashfree, but there's no luck, I'd to switch back to original 5003...

I assume we need a tool that fake the digital sign, this BIOS must be protected against non-official ROMs (maybe to avoid malicious software which can infect BIOS, what acually is pretty fine)

**Comment by zgrcn41** [comment]
rom file does not match existing bios size

**Reply by SpontaneousJ** [reply]
I got this same error unfortunately.

**Reply by BanefulDemon** [reply]
Same error, Z390-F

**Comment by zgrcn41** [comment]
Error please help

**Comment by zgrcn41** [comment]
asus pb450ma2

**Comment by NOHEATED** [comment]
UEFI DXE driver status: Not loaded (0x000000000000000A)

Still doesn`t work

Motherboard: Asus TUF B450M-PRO GAMING

ReBar in BIOS set on auto

> DXE Driver configuration:
>         isDirty:           0
>         OptionFlags:       0x0002
>                                - nGlobalEnable:      2
>                                - skipS3Resume:       0
>                                - overrideBarSize:    0
>                                - hasSetupVarCRC:     0
>                                - disableSetupVarCRC: 0
>         SetupVarCRC:       0x0000'0000'0000'0000
>         nPciBarSize:       0
>         nGPUSelectorCount: 0
>         nGPUConfigCount:   1
>                 GPUConfig1:    deviceID:        1E84
>                 GPUConfig1:    subsysVendorID:  1458
>                 GPUConfig1:    subsysDeviceID:  4008
>                 GPUConfig1:    bus:             08
>                 GPUConfig1:    device:          00
>                 GPUConfig1:    function:        0
>                 GPUConfig1:    BAR0 base:       0xF500'0000
>                 GPUConfig1:    BAR0 top:        0xF5FF'FFFF
> 
>         nBridgeCount:      1
>                 BridgeConfig1: vendorID:        1022
>                 BridgeConfig1: deviceID:        1483
>                 BridgeConfig1: bus:             00
>                 BridgeConfig1: device:          03
>                 BridgeConfig1: function:        1
>                 BridgeConfig1: secondary bus:   08

![gpu-z](https://github.com/terminatorul/NvStrapsReBar/assets/21317070/42c567b2-1bc5-460e-b7e4-f2f3040f734a)

**Reply by hentzchenklein** [reply]
Same here on x299 Prime AII

**Comment by Artasdmc** [comment]
Same here. B450F.

I don't think BIOS is actually flashing even with the tool and .rom conversion.
Flashing BIOS with EZ-Flash just gives an error that it's not proper BIOS, as it probably requires signature.

I guess Asus owners are not getting this on their boards with older GPUs.

**Reply by Artasdmc** [reply]
Getting back at this saying I've even bricked my motherboard by trying to downgrade BIOS with AFUDOS, as Ez-flash does not let downgrade and mobo doesn't have a flashback feature. 

Getting a programmer sent in tomorrow to reflash the chip physically with modified bios. I'll post an update if it works, or only stock untouched bios can be used.

**Reply by Artasdmc** [reply]
Forgot to get back. Physically flashed modified bios with a programmer and it doesn't work still on B450F. Tried 2 different placements.

I guess for people with Asus boards, I would just not bother with this. You'll end up bricking a board if you do try and downgrade bios forcefully.

**Reply by NOHEATED** [reply]
The worst part that it did worked once just fine the regular way, but it doesn`t work anymore

**Comment by stupable** [comment]
x570i strix ,getting error 22, tried asus ai suite trick, also did not work
i got this far tho...
![Untitled-2](https://github.com/terminatorul/NvStrapsReBar/assets/167392934/3e80079a-7a6c-4926-85c5-2216927ef481)
![3](https://github.com/terminatorul/NvStrapsReBar/assets/167392934/434001f6-9d19-4444-99ae-224e4c93047a)

**Comment by DeadGames2** [comment]
hey i may be retarted
in the UEFI tool i should find SUBType AMI Aptio or in text AMI Aptio or you mean AMI Aptio as a Intel Image etc.
yeah im retarted dont ask, would be happy for help tho

**Comment by wavexLINE** [comment]
I have also encountered "UEFI DXE driver status: Not loaded (0x000000000000000A)"

I followed these steps to a T. This was actually my fallback plan as I couldn't get updating from EZ Flash to accept my modified bios.

All files contained the proper UEFI DXE components along the way. I restored bios defaults before starting the process. Booting from non-UEFI USB wasn't straightforward as I had to restart after enabling CSM for the drive to show up as non-UEFI before booting from it. The auto run from USB completed as normal. Not sure what to make of it.

System specs:
CPU: 5800X3D
Mobo: Crosshair VI Hero X370 (Bios 8702, both as default and as modified with UEFI tool)
RTX 2060

Do I need to revert to an older BIOS for this to work, as I read in a different forum? I'm not sure if this is user error or incompatibility or a simple mistake.

**Comment by Bakashii** [comment]
Hey thank you so much! Your guide helped me get NvStrapsReBar installed.

**Comment by AnthonyZ1993** [comment]
I followed the steps step by step, after having updated the bios with the words done as shown in the image, I restarted and now I find that the motherboard cannot start it with an orange LED on a fixed Dram. I can no longer find a way to start it

I've made several attempts like

remove the battery
reset cmos pins
remove the rams 

no results

asus z390e gaming 

Is there a solution to go back?

**Comment by Artasdmc** [comment]
You will have to buy a programmer and flash bios manually by jumping pins
on bios chip. This happened to me too.

**Reply by AnthonyZ1993** [reply]
Could you explain to me in detail how to do it? Where to buy the programmer??  email: anthony1993ct@gmail.com I hope you can answer us, following this guide ruined my day, I can't use my Z390 and gaming... don't update this way or you will find your motherboard fried.

**Comment by Artasdmc** [comment]
Google CH340 programmer.
Get one that comes with 3v adapter and a pair of clips.

Once you have it, download your original bios from manufacturer. Download
one of many flashing tools for CH340.

Then use clips to connect the programmer on to bios chip and flash original
bios manually with CH340.

There's plenty of YouTube videos on this.

**Comment by Artasdmc** [comment]
You'll just need a separate computer to do this. Laptop if you have one.

**Comment by Joooooooooooooooooe** [comment]
In the "you should eventually find the region that contains a lot of DXE drivers" part, I didn't see "a lot of  DXE drivers", instead only see one DXE driver, I don't know if this is fine, have already checked other parts dont include any DXE drivers.
But I still follow the whole steps, there is no obvious error occured, and it seems not working, still got "UEFI DXE driver status: Not loaded (0x000000000000000A)".
ASUS strix z390-e gaming.

**Comment by NoOneBlood** [comment]
ROG STRIX Z370-F and GTX 1660ti are working for me. Thanks!

**Comment by maCro532** [comment]
It works on ROG STRIX Z370-H and RTX 2080 ti. Thanks

**Comment by Misticc1** [comment]
for "insert after" UEFITool 0.21.5 version is required.

**Comment by Drift91** [comment]
Unfortunately I could never get AFUDOS or AFUWin64 to work. They'd gladly brick my BIOS but would never add the mod to it. I suspect that the BIOS has some kind of safeguard to detect when you use such tools to flash a modded BIOS and deletes the mods. After AFUWin64 bricked my BIOS a second time, I had no choice but to break out the CH341A again; this time I decided to flash the modded BIOS, and as much as I despise using the SPI flash programmer, it worked perfectly to flash the modded version.

**Reply by chris1892006** [reply]
Hi, im on x570i gaming board here and want to flash modded bios.  Will I have any issues?
![asus_rog_strix_x570_i_gaming_pcb](https://github.com/user-attachments/assets/2a4ee535-2787-4569-a342-ae15e6a8d8d2)

**Reply by Koekieezz** [reply]
> x570i gaming

@chris1892006 You would need to downgrade to bios before agesa 1.2.x.x, which means you might need to use old cpu like 1000 series up to 3000 series. then use flashrom/flashprog DOS version to flash the modified bios.

I do recommend to Update your bios to the latest bios, boot to flashrom/flashprog DOS, and dump your bios, then proceed to make a copy of your dumped bios, insert the .ffs with UEFITool **v0.25.0** (DO NOT USE ANY VERSION ABOVE, IT COULD CORRUPT PAD FILES), save the bios, then proceed to downgrade your bios to older version before agesa 1.2.x.x, once it is done, use flashrom/flashprog DOS again to flash your dump-edited bios.

you can learn about how to use flashrom/flashprog DOS version [here](https://winraid.level1techs.com/t/tool-flashprog-1-1-flashrom-v1-5-dos/102796?u=koekieezz).
